In [1]:
import pandas as pd
f500 = pd.read_csv('../Dataset/f500.csv', index_col = 0) # index_col specifies what column use to index the row axis
f500.index.name = None # removes the name of index axis
f500.head(5)

rank  revenues  revenue_change  profits  assets  \
Walmart                      1    485873             0.8  13643.0  198825   
State Grid                   2    315199            -4.4   9571.3  489838   
Sinopec Group                3    267518            -9.1   1257.9  310726   
China National Petroleum     4    262573           -12.3   1867.5  585619   
Toyota Motor                 5    254694             7.7  16899.3  437575   

                          profit_change                  ceo  \
Walmart                            -7.2  C. Douglas McMillon   
State Grid                         -6.2              Kou Wei   
Sinopec Group                     -65.0            Wang Yupu   
China National Petroleum          -73.7        Zhang Jianhua   
Toyota Motor                      -12.3          Akio Toyoda   

                                          industry                  sector  \
Walmart                      General Merchandisers               Retailing   
State Grid                               Utilities                  Energy   
Sinopec Group                   Petroleum Refining                  Energy   
China National Petroleum        Petroleum Refining                  Energy   
Toyota Motor              Motor Vehicles and Parts  Motor Vehicles & Parts   

                          previous_rank country      hq_location  \
Walmart                               1     USA  Bentonville, AR   
State Grid                            2   China   Beijing, China   
Sinopec Group                         4   China   Beijing, China   
China National Petroleum              3   China   Beijing, China   
Toyota Motor                          8   Japan    Toyota, Japan   

                                               website  \
Walmart                         http://www.walmart.com   
State Grid                      http://www.sgcc.com.cn   
Sinopec Group                   http://www.sinopec.com   
China National Petroleum        http://www.cnpc.com.cn   
Toyota Motor              http://www.toyota-global.com   

                          years_on_global_500_list  employees  \
Walmart                                         23    2300000   
State Grid                                      17     926067   
Sinopec Group                                   19     713288   
China National Petroleum                        17    1512048   
Toyota Motor                                    23     364445   

                          total_stockholder_equity  
Walmart                                      77798  
State Grid                                  209456  
Sinopec Group                               106523  
China National Petroleum                    301893  
Toyota Motor                                157210

In [2]:
industry = f500['industry'] # selecting a single column in a dataframe
industry

Walmart                                     General Merchandisers
State Grid                                              Utilities
Sinopec Group                                  Petroleum Refining
China National Petroleum                       Petroleum Refining
Toyota Motor                             Motor Vehicles and Parts
                                               ...               
Teva Pharmaceutical Industries                    Pharmaceuticals
New China Life Insurance          Insurance: Life, Health (stock)
Wm. Morrison Supermarkets                    Food and Drug Stores
TUI                                               Travel Services
AutoNation                                    Specialty Retailers
Name: industry, Length: 500, dtype: object

In [3]:
country = f500['country'] # selecting a specific column in a dataframe
country

Walmart                               USA
State Grid                          China
Sinopec Group                       China
China National Petroleum            China
Toyota Motor                        Japan
                                   ...   
Teva Pharmaceutical Industries     Israel
New China Life Insurance            China
Wm. Morrison Supermarkets         Britain
TUI                               Germany
AutoNation                            USA
Name: country, Length: 500, dtype: object

In [4]:
revenue_years = f500[['revenues', 'years_on_global_500_list']] # selecting a list of specific columns in a dataframe
revenue_years

revenues  years_on_global_500_list
Walmart                           485873                        23
State Grid                        315199                        17
Sinopec Group                     267518                        19
China National Petroleum          262573                        17
Toyota Motor                      254694                        23
...                                  ...                       ...
Teva Pharmaceutical Industries     21903                         1
New China Life Insurance           21796                         2
Wm. Morrison Supermarkets          21741                        13
TUI                                21655                        23
AutoNation                         21609                        12

[500 rows x 2 columns]

In [5]:
ceo_to_sector = f500.loc[:, 'ceo':'sector'] # selecting a slice of columns in dataframe
ceo_to_sector

ceo  \
Walmart                         C. Douglas McMillon   
State Grid                                  Kou Wei   
Sinopec Group                             Wang Yupu   
China National Petroleum              Zhang Jianhua   
Toyota Motor                            Akio Toyoda   
...                                             ...   
Teva Pharmaceutical Industries    Yitzhak Peterburg   
New China Life Insurance                   Wan Feng   
Wm. Morrison Supermarkets            David T. Potts   
TUI                               Friedrich Joussen   
AutoNation                       Michael J. Jackson   

                                                       industry  \
Walmart                                   General Merchandisers   
State Grid                                            Utilities   
Sinopec Group                                Petroleum Refining   
China National Petroleum                     Petroleum Refining   
Toyota Motor                           Motor Vehicles and Parts   
...                                                         ...   
Teva Pharmaceutical Industries                  Pharmaceuticals   
New China Life Insurance        Insurance: Life, Health (stock)   
Wm. Morrison Supermarkets                  Food and Drug Stores   
TUI                                             Travel Services   
AutoNation                                  Specialty Retailers   

                                                sector  
Walmart                                      Retailing  
State Grid                                      Energy  
Sinopec Group                                   Energy  
China National Petroleum                        Energy  
Toyota Motor                    Motor Vehicles & Parts  
...                                                ...  
Teva Pharmaceutical Industries             Health Care  
New China Life Insurance                    Financials  
Wm. Morrison Supermarkets           Food & Drug Stores  
TUI                                  Business Services  
AutoNation                                   Retailing  

[500 rows x 3 columns]

In [6]:
toyota = f500.loc['Toyota Motor'] # selecting a single row in a dataframe
toyota

rank                                                   5
revenues                                          254694
revenue_change                                       7.7
profits                                          16899.3
assets                                            437575
profit_change                                      -12.3
ceo                                          Akio Toyoda
industry                        Motor Vehicles and Parts
sector                            Motor Vehicles & Parts
previous_rank                                          8
country                                            Japan
hq_location                                Toyota, Japan
website                     http://www.toyota-global.com
years_on_global_500_list                              23
employees                                         364445
total_stockholder_equity                          157210
Name: Toyota Motor, dtype: object

In [7]:
drink_companies = f500.loc[['Anheuser-Busch InBev', 'Coca-Cola', 'Heineken Holding']] # Selecting list of unique rows in a dataframe
drink_companies

rank  revenues  revenue_change  profits  assets  \
Anheuser-Busch InBev   206     45905             5.3   1241.0  258381   
Coca-Cola              235     41863            -5.5   6527.0   87270   
Heineken Holding       468     23044            -0.7    861.5   41469   

                      profit_change                        ceo   industry  \
Anheuser-Busch InBev          -85.0               Carlos Brito  Beverages   
Coca-Cola                     -11.2           James B. Quincey  Beverages   
Heineken Holding              -18.9  Jean-Francois van Boxmeer  Beverages   

                                         sector  previous_rank      country  \
Anheuser-Busch InBev  Food, Beverages & Tobacco            211      Belgium   
Coca-Cola             Food, Beverages & Tobacco            206          USA   
Heineken Holding      Food, Beverages & Tobacco            459  Netherlands   

                                 hq_location  \
Anheuser-Busch InBev         Leuven, Belgium   
Coca-Cola                        Atlanta, GA   
Heineken Holding      Amsterdam, Netherlands   

                                                website  \
Anheuser-Busch InBev            http://www.ab-inbev.com   
Coca-Cola               http://www.coca-colacompany.com   
Heineken Holding      http://www.theheinekencompany.com   

                      years_on_global_500_list  employees  \
Anheuser-Busch InBev                        12     206633   
Coca-Cola                                   23     100300   
Heineken Holding                            11      73525   

                      total_stockholder_equity  
Anheuser-Busch InBev                     71339  
Coca-Cola                                23062  
Heineken Holding                          6958

In [8]:
middle_companies = f500['Tata Motors' : 'Nationwide'] # Selecting a slice of rows in a dataframe
middle_companies

rank  revenues  revenue_change  profits  \
Tata Motors                       247     40329            -4.2   1111.6   
Aluminum Corp. of China           248     40278             6.0   -282.5   
Mitsui                            249     40275             1.6   2825.3   
Manulife Financial                250     40238            49.4   2209.7   
China Minsheng Banking            251     40234            -5.2   7201.6   
China Pacific Insurance (Group)   252     40193             2.2   1814.9   
American Airlines Group           253     40180            -2.0   2676.0   
Nationwide                        254     40074            -0.4    334.3   

                                 assets  profit_change                   ceo  \
Tata Motors                       42162          -34.0      Guenter Butschek   
Aluminum Corp. of China           75089            NaN              Yu Dehui   
Mitsui                           103231            NaN       Tatsuo Yasunaga   
Manulife Financial               537461           28.9     Donald A. Guloien   
China Minsheng Banking           848389           -1.8         Zheng Wanchun   
China Pacific Insurance (Group)  146873          -35.7          Huo Lianhong   
American Airlines Group           51274          -64.8     W. Douglas Parker   
Nationwide                       197790          -42.4  Stephen S. Rasmussen   

                                                                  industry  \
Tata Motors                                       Motor Vehicles and Parts   
Aluminum Corp. of China                                             Metals   
Mitsui                                                             Trading   
Manulife Financial                         Insurance: Life, Health (stock)   
China Minsheng Banking                       Banks: Commercial and Savings   
China Pacific Insurance (Group)            Insurance: Life, Health (stock)   
American Airlines Group                                           Airlines   
Nationwide                       Insurance: Property and Casualty (Mutual)   

                                                 sector  previous_rank  \
Tata Motors                      Motor Vehicles & Parts            226   
Aluminum Corp. of China                       Materials            262   
Mitsui                                      Wholesalers            245   
Manulife Financial                           Financials            394   
China Minsheng Banking                       Financials            221   
China Pacific Insurance (Group)              Financials            251   
American Airlines Group                  Transportation            236   
Nationwide                                   Financials            241   

                                country               hq_location  \
Tata Motors                       India             Mumbai, India   
Aluminum Corp. of China           China            Beijing, China   
Mitsui                            Japan              Tokyo, Japan   
Manulife Financial               Canada  Toronto, Ontario, Canada   
China Minsheng Banking            China            Beijing, China   
China Pacific Insurance (Group)   China           Shanghai, China   
American Airlines Group             USA            Fort Worth, TX   
Nationwide                          USA              Columbus, OH   

                                                   website  \
Tata Motors                      http://www.tatamotors.com   
Aluminum Corp. of China           http://www.chalco.com.cn   
Mitsui                               http://www.mitsui.com   
Manulife Financial                 http://www.manulife.com   
China Minsheng Banking              http://www.cmbc.com.cn   
China Pacific Insurance (Group)     http://www.cpic.com.cn   
American Airlines Group                  http://www.aa.com   
Nationwide                       http://www.nationwide.com   

                                 years_on_global_500_list  employees  \
Tata

In [9]:
country_count = country.value_counts() # since the country data is a series,
#we can use the value_count function to count the number of times each unique value appears in the series
country_count

USA             132
China           109
Japan            51
Germany          29
France           29
Britain          24
South Korea      15
Switzerland      14
Netherlands      14
Canada           11
Spain             9
India             7
Italy             7
Australia         7
Brazil            7
Taiwan            6
Russia            4
Ireland           4
Sweden            3
Singapore         3
Mexico            2
Venezuela         1
Israel            1
Saudi Arabia      1
Thailand          1
Luxembourg        1
U.A.E             1
Finland           1
Norway            1
Indonesia         1
Denmark           1
Malaysia          1
Turkey            1
Belgium           1
Name: country, dtype: int64

In [10]:
india = country_count['India'] # selecting a single item from a series
india

7

In [11]:
north_america = country_count[['USA', 'Canada', 'Mexico']] # selecting unique items from a series
north_america

USA       132
Canada     11
Mexico      2
Name: country, dtype: int64

In [12]:
spain_to_russia = country_count['Spain':'Russia'] # selecting a slice of countries from a series
spain_to_russia

Spain        9
India        7
Italy        7
Australia    7
Brazil       7
Taiwan       6
Russia       4
Name: country, dtype: int64

In [13]:
big_movers = f500.loc[['Aviva', 'HP', 'JD.com', 'BHP Billiton'], ['rank', 'previous_rank']] 
# selecting multiple rows and multiple columns
big_movers

rank  previous_rank
Aviva           90            279
HP             194             48
JD.com         261            366
BHP Billiton   350            168

In [14]:
bottom_companies = f500.loc['National Grid':'AutoNation', ['rank', 'sector', 'country']] #selecting a slice of rows and multiple columns
bottom_companies

rank              sector  country
National Grid                           491              Energy  Britain
Dollar General                          492           Retailing      USA
Telecom Italia                          493  Telecommunications    Italy
Xiamen ITG Holding Group                494         Wholesalers    China
Xinjiang Guanghui Industry Investment   495         Wholesalers    China
Teva Pharmaceutical Industries          496         Health Care   Israel
New China Life Insurance                497          Financials    China
Wm. Morrison Supermarkets               498  Food & Drug Stores  Britain
TUI                                     499   Business Services  Germany
AutoNation                              500           Retailing      USA

In [15]:
rank_change = f500['previous_rank'] - f500['rank'] # calculating the change in companies ranks
rank_change

Walmart                             0
State Grid                          0
Sinopec Group                       1
China National Petroleum           -1
Toyota Motor                        3
                                 ... 
Teva Pharmaceutical Industries   -496
New China Life Insurance          -70
Wm. Morrison Supermarkets         -61
TUI                               -32
AutoNation                       -500
Length: 500, dtype: int64

In [16]:
rank_change_max = rank_change.max() # calculating the maximum change in companies ranks
rank_change_max

226

In [17]:
rank_change_min = rank_change.min() # calculating the minimum change in companies ranks
rank_change_min

-500

### A Brief Explanation of some error
According to the data dictionary, this list should only rank companies on a scale of 1 to 500. Even if the company ranked 1st in the previous year moved to 500th this year, the rank change calculated would be -499. This indicates that there is incorrect data in either the rank column or previous_rank column.

In [18]:
rank_change.describe()

count    500.000000
mean     -28.366000
std      108.602823
min     -500.000000
25%      -28.250000
50%       -4.000000
75%        8.250000
max      226.000000
dtype: float64

In [19]:
f500.describe() # this shows the descriptives for all columns in the dataframe with float and integer values
# f500.describe(include=['O']) To show the descriptives for object columns we use the code 

rank       revenues  revenue_change       profits        assets  \
count  500.000000     500.000000      498.000000    499.000000  5.000000e+02   
mean   250.500000   55416.358000        4.538353   3055.203206  2.436323e+05   
std    144.481833   45725.478963       28.549067   5171.981071  4.851937e+05   
min      1.000000   21609.000000      -67.300000 -13038.000000  3.717000e+03   
25%    125.750000   29003.000000       -5.900000    556.950000  3.658850e+04   
50%    250.500000   40236.000000        0.550000   1761.600000  7.326150e+04   
75%    375.250000   63926.750000        6.975000   3954.000000  1.805640e+05   
max    500.000000  485873.000000      442.300000  45687.000000  3.473238e+06   

       profit_change  previous_rank  years_on_global_500_list     employees  \
count     436.000000     500.000000                500.000000  5.000000e+02   
mean       24.152752     222.134000                 15.036000  1.339983e+05   
std       437.509566     146.941961                  7.932752  1.700878e+05   
min      -793.700000       0.000000                  1.000000  3.280000e+02   
25%       -22.775000      92.750000                  7.000000  4.293250e+04   
50%        -0.350000     219.500000                 17.000000  9.291050e+04   
75%        17.700000     347.250000                 23.000000  1.689172e+05   
max      8909.500000     500.000000                 23.000000  2.300000e+06   

       total_stockholder_equity  
count                500.000000  
mean               30628.076000  
std                43642.576833  
min               -59909.000000  
25%                 7553.750000  
50%                15809.500000  
75%                37828.500000  
max               301893.000000

The assets and employees  column shows values that are too long to display neatly e.g (5.000000e+02) these values hence pandas displays them in a special kind of scientific notation (the e notation). This results in these values being calculated as (5 * 10 ** 2 for 5.000000e+02 and 2.436323 * 10 ** 5 for 2.436323e+05)

In [20]:
country.describe() # none float or integer values returns a different kind of statistics

count     500
unique     34
top       USA
freq      132
Name: country, dtype: object

In [21]:
rank = f500['rank']
rank.describe()

count    500.000000
mean     250.500000
std      144.481833
min        1.000000
25%      125.750000
50%      250.500000
75%      375.250000
max      500.000000
Name: rank, dtype: float64

In [22]:
previous_rank = f500['previous_rank']
previous_rank.describe()

count    500.000000
mean     222.134000
std      146.941961
min        0.000000
25%       92.750000
50%      219.500000
75%      347.250000
max      500.000000
Name: previous_rank, dtype: float64

The above data shows that the minimum value in the `previous_rank` column is zero. This makes no sense as companies can only be ranked from 1 to 500, so the value zero is an error. 
Using *method chaning* we can count the number of times the number zero appeared in the `previous_rank` column

In [23]:
previous_rank.value_counts().loc[0]

33

we confirmed that 33 companies in the dataframe have a value of 0 in the `previous_rank` column. Given that multiple companies have a 0 rank, we might conclude that these companies didn't have a rank at all for the previous year. It would make more sense for us to replace these values with a *null value* instead.
We need to explore the rest of our dataframe to make sure there are no other data issues.

In [24]:
max_f500 = f500.max(numeric_only = True) # Showing the maximum values for only numeric columns in the dataframe
max_f500

rank                            500.0
revenues                     485873.0
revenue_change                  442.3
profits                       45687.0
assets                      3473238.0
profit_change                  8909.5
previous_rank                   500.0
years_on_global_500_list         23.0
employees                   2300000.0
total_stockholder_equity     301893.0
dtype: float64

In [25]:
min_f500 = f500.min(numeric_only = True)
min_f500

rank                            1.0
revenues                    21609.0
revenue_change                -67.3
profits                    -13038.0
assets                       3717.0
profit_change                -793.7
previous_rank                   0.0
years_on_global_500_list        1.0
employees                     328.0
total_stockholder_equity   -59909.0
dtype: float64

The above code shows that apart from the `previous_rank` column all other columns in our dataframe looks quite reasonable. 
Previously, we concluded that companies with a rank of zero didn't have a rank at all. Next, we'll replace these values with a null value to clearly indicate that the value is missing. `np.nan` is used in pandas to represent values that can't be represented numerically, most commonly missing values.

In [26]:
import numpy as np
f500.loc[f500["previous_rank"] == 0, 'previous_rank'] = np.nan #Replacing all zero values in previous_rank with NaN values indicating that the value is missing

In [27]:
f500.tail()

rank  revenues  revenue_change  profits  \
Teva Pharmaceutical Industries   496     21903            11.5    329.0   
New China Life Insurance         497     21796           -13.3    743.9   
Wm. Morrison Supermarkets        498     21741           -11.3    406.4   
TUI                              499     21655            -5.5   1151.7   
AutoNation                       500     21609             3.6    430.5   

                                assets  profit_change                 ceo  \
Teva Pharmaceutical Industries   92890          -79.3   Yitzhak Peterburg   
New China Life Insurance        100609          -45.6            Wan Feng   
Wm. Morrison Supermarkets        11630           20.4      David T. Potts   
TUI                              16247          195.5   Friedrich Joussen   
AutoNation                       10060           -2.7  Michael J. Jackson   

                                                       industry  \
Teva Pharmaceutical Industries                  Pharmaceuticals   
New China Life Insurance        Insurance: Life, Health (stock)   
Wm. Morrison Supermarkets                  Food and Drug Stores   
TUI                                             Travel Services   
AutoNation                                  Specialty Retailers   

                                            sector  previous_rank  country  \
Teva Pharmaceutical Industries         Health Care            NaN   Israel   
New China Life Insurance                Financials          427.0    China   
Wm. Morrison Supermarkets       Food & Drug Stores          437.0  Britain   
TUI                              Business Services          467.0  Germany   
AutoNation                               Retailing            NaN      USA   

                                         hq_location  \
Teva Pharmaceutical Industries  Petach Tikva, Israel   
New China Life Insurance              Beijing, China   
Wm. Morrison Supermarkets          Bradford, Britain   
TUI                                 Hanover, Germany   
AutoNation                       Fort Lauderdale, FL   

                                                    website  \
Teva Pharmaceutical Industries     http://www.tevapharm.com   
New China Life Insurance        http://www.newchinalife.com   
Wm. Morrison Supermarkets          http://www.morrisons.com   
TUI                                 http://www.tuigroup.com   
AutoNation                        http://www.autonation.com   

                                years_on_global_500_list  employees  \
Teva Pharmaceutical Industries                         1      56960   
New China Life Insurance                               2      54378   
Wm. Morrison Supermarkets                             13      77210   
TUI                                                   23      66779   
AutoNation                                            12      26000   

                                total_stockholder_equity  
Teva Pharmaceutical Industries                     33337  
New China Life Insurance                            8507  
Wm. Morrison Supermarkets                           5111  
TUI                                                 3006  
AutoNation                                          2310

The above table shows that the values that were previously zero have been updated to NaN values.

Now that we've corrected the data, let's create the rank_change series again. This time, we'll add it to our f500 dataframe as a new column.


In [28]:
f500['rank_change'] = f500['previous_rank'] - f500['rank'] # Assigning values to a non existent column creates a new column in the dataframe
f500['rank_change']

Walmart                            0.0
State Grid                         0.0
Sinopec Group                      1.0
China National Petroleum          -1.0
Toyota Motor                       3.0
                                  ... 
Teva Pharmaceutical Industries     NaN
New China Life Insurance         -70.0
Wm. Morrison Supermarkets        -61.0
TUI                              -32.0
AutoNation                         NaN
Name: rank_change, Length: 500, dtype: float64

In [29]:
f500['rank_change'].describe()

count    467.000000
mean      -3.533191
std       44.293603
min     -199.000000
25%      -21.000000
50%       -2.000000
75%       10.000000
max      226.000000
Name: rank_change, dtype: float64

In [30]:
f500['industry'][f500['country'] == 'USA'].value_counts().head(2) # This shows the top two industries in the USA 

Banks: Commercial and Savings               8
Insurance: Property and Casualty (Stock)    7
Name: industry, dtype: int64

### The table below summarizes how we can use DataFrame.iloc[] and Series.iloc[] to select by integer position:

|Select by integer position|	 Explicit Syntax|    Shorthand Convention|
|:---------------------------- | :----------------:| -------------------: |
|Single column from dataframe|    df.iloc[:,3]|	
|List of columns from dataframe|  df.iloc[:,[3,5,6]]|
|Slice of columns from dataframe| df.iloc[:,3:7]|
|Single row from dataframe|	      df.iloc[20]|
|List of rows from dataframe|	  df.iloc[[0,3,8]]|	
|Slice of rows from dataframe|	  df.iloc[3:5]|	       **df** [3:5]|
|Single items from series|	      s.iloc[8]	 |         **s** [8]|
|List of item from series|	      s.iloc[[2,8,1]]|	   **s** [[2,8,1]]|
|Slice of items from series	|     s.iloc[5:10]	 |     **s** [5:10]|

In [31]:
null_revenues = f500['revenue_change'].isnull() # this checks if there are any null values in the revenues_change column
null_revenues.head() # returns the first five dataset for the revenue_change column

Walmart                     False
State Grid                  False
Sinopec Group               False
China National Petroleum    False
Toyota Motor                False
Name: revenue_change, dtype: bool

In [32]:
company_with_null_revenue_change = f500[null_revenues][['sector', 'country', 'revenues', 'revenue_change']] # The null_revenue
# series is used to filter the dataset with some selected columns
company_with_null_revenue_change

sector  country  revenues  revenue_change
Uniper                          Energy  Germany     74407             NaN
Hewlett Packard Enterprise  Technology      USA     50123             NaN

Above we see that there are two companies (Uniper and Hewlett Packard (HP) enterprise) in the entire dataset whose revenue change is null i.e not a number. And these companies fall in the `energy` and `technology` sectors and they are located in Germany and USA 

In [33]:
null_previous = f500['previous_rank'].isnull()
null_previous_rank = f500[null_previous][['rank', 'previous_rank']] 
# Above we create a new dataframe (null_previous_rank) that indexes the entire dataframe(f500) that meets the criteria
# of null_previous, but return only two columns
# The code can be run in the reverse order with column specification coming before the row indexing
null_previous_rank

rank  previous_rank
Legal & General Group                    49            NaN
Uniper                                   91            NaN
Dell Technologies                       124            NaN
Anbang Insurance Group                  139            NaN
Albertsons Cos.                         141            NaN
Hewlett Packard Enterprise              181            NaN
Hengli Group                            268            NaN
Johnson Controls International          272            NaN
Chubb                                   342            NaN
Charter Communications                  376            NaN
Medtronic                               377            NaN
Facebook                                393            NaN
Kraft Heinz                             410            NaN
Nokia                                   415            NaN
Standard Life                           432            NaN
Brookfield Asset Management             441            NaN
Mercantil Servicios Financieros         442            NaN
Samsung C&T                             447            NaN
Heraeus Holding                         457            NaN
Yango Financial Holding                 459            NaN
Alibaba Group Holding                   462            NaN
Progressive                             464            NaN
Country Garden Holdings                 467            NaN
Altice                                  473            NaN
Tencent Holdings                        478            NaN
Aflac                                   483            NaN
Suning Commerce Group                   485            NaN
Xiamen C&D                              488            NaN
Dollar General                          492            NaN
Xiamen ITG Holding Group                494            NaN
Xinjiang Guanghui Industry Investment   495            NaN
Teva Pharmaceutical Industries          496            NaN
AutoNation                              500            NaN

###### The Above output shows the companies whose `previous_rank` column is null and also the current rank of those companies

In [34]:
null_previous_rank.count()

rank             33
previous_rank     0
dtype: int64

The output above shows that there are 33 such companies with null `previous_rank` values

In [35]:
null_previous_rank.iloc[0:5] # Selecting the first five columns for the null_previous_rank dataframe

rank  previous_rank
Legal & General Group     49            NaN
Uniper                    91            NaN
Dell Technologies        124            NaN
Anbang Insurance Group   139            NaN
Albertsons Cos.          141            NaN

In [36]:
previously_ranked = f500[f500['previous_rank'].notnull()] # create a new dataframe (previously_ranked) that 
#selects all rows whoseprevious_rank column is not null
previously_ranked

rank  revenues  revenue_change  profits  assets  \
Walmart                       1    485873             0.8  13643.0  198825   
State Grid                    2    315199            -4.4   9571.3  489838   
Sinopec Group                 3    267518            -9.1   1257.9  310726   
China National Petroleum      4    262573           -12.3   1867.5  585619   
Toyota Motor                  5    254694             7.7  16899.3  437575   
...                         ...       ...             ...      ...     ...   
National Grid               491     22036            -3.2  10150.6   82310   
Telecom Italia              493     21941           -17.4   1999.4   74295   
New China Life Insurance    497     21796           -13.3    743.9  100609   
Wm. Morrison Supermarkets   498     21741           -11.3    406.4   11630   
TUI                         499     21655            -5.5   1151.7   16247   

                           profit_change                  ceo  \
Walmart                             -7.2  C. Douglas McMillon   
State Grid                          -6.2              Kou Wei   
Sinopec Group                      -65.0            Wang Yupu   
China National Petroleum           -73.7        Zhang Jianhua   
Toyota Motor                       -12.3          Akio Toyoda   
...                                  ...                  ...   
National Grid                      160.2       John Pettigrew   
Telecom Italia                       NaN      Flavio Cattaneo   
New China Life Insurance           -45.6             Wan Feng   
Wm. Morrison Supermarkets           20.4       David T. Potts   
TUI                                195.5    Friedrich Joussen   

                                                  industry  \
Walmart                              General Merchandisers   
State Grid                                       Utilities   
Sinopec Group                           Petroleum Refining   
China National Petroleum                Petroleum Refining   
Toyota Motor                      Motor Vehicles and Parts   
...                                                    ...   
National Grid                                    Utilities   
Telecom Italia                          Telecommunications   
New China Life Insurance   Insurance: Life, Health (stock)   
Wm. Morrison Supermarkets             Food and Drug Stores   
TUI                                        Travel Services   

                                           sector  previous_rank  country  \
Walmart                                 Retailing            1.0      USA   
State Grid                                 Energy            2.0    China   
Sinopec Group                              Energy            4.0    China   
China National Petroleum                   Energy            3.0    China   
Toyota Motor               Motor Vehicles & Parts            8.0    Japan   
...                                           ...            ...      ...   
National Grid                              Energy          471.0  Britain   
Telecom Italia                 Telecommunications          404.0    Italy   
New China Life Insurance               Financials          427.0    China   
Wm. Morrison Supermarkets      Food & Drug Stores          437.0  Britain   
TUI                             Business Services          467.0  Germany   

                                 hq_location                       website  \
Walmart                      Bentonville, AR        http://www.walmart.com   
State Grid                    Beijing, China        http://www.sgcc.com.cn   
Sinopec Group                 Beijing, China        http://www.sinopec.com   
China National Petroleum      Beijing, China        http://www.cnpc.com.cn   
Toyota Motor                   Toyota, Japan  http://www.toyota-global.com   
...                                      ...                           ...   
National Grid                London, Britain   http://www.nationalgrid.com   
Telecom Italia 

We notice that the number of rows in the `previously_ranked` dataframe (467) is smaller than what is obtained in the f500 dataframe (500). This is because `previously_ranked` only picked values that are `notnull` from f500

In [37]:
rank_change = previously_ranked['previous_rank'] - previously_ranked['rank']
rank_change

Walmart                       0.0
State Grid                    0.0
Sinopec Group                 1.0
China National Petroleum     -1.0
Toyota Motor                  3.0
                             ... 
National Grid               -20.0
Telecom Italia              -89.0
New China Life Insurance    -70.0
Wm. Morrison Supermarkets   -61.0
TUI                         -32.0
Length: 467, dtype: float64

The `rank_change` column from the `previously_ranked` dataframe also also has smaller amount of rows (467) 

In [38]:
no_change = (rank_change == 0)
q = previously_ranked[no_change] [['industry','country']]
q

industry  country
Walmart                           General Merchandisers      USA
State Grid                                    Utilities    China
Apple                       Computers, Office Equipment      USA
Ford Motor                     Motor Vehicles and Parts      USA
Allianz        Insurance: Property and Casualty (Stock)  Germany
Peugeot                        Motor Vehicles and Parts   France
Continental                    Motor Vehicles and Parts  Germany
Compass Group                             Food Services  Britain

The above output shows the companies that maintained the same ranking 

In [39]:
previously_ranked

rank  revenues  revenue_change  profits  assets  \
Walmart                       1    485873             0.8  13643.0  198825   
State Grid                    2    315199            -4.4   9571.3  489838   
Sinopec Group                 3    267518            -9.1   1257.9  310726   
China National Petroleum      4    262573           -12.3   1867.5  585619   
Toyota Motor                  5    254694             7.7  16899.3  437575   
...                         ...       ...             ...      ...     ...   
National Grid               491     22036            -3.2  10150.6   82310   
Telecom Italia              493     21941           -17.4   1999.4   74295   
New China Life Insurance    497     21796           -13.3    743.9  100609   
Wm. Morrison Supermarkets   498     21741           -11.3    406.4   11630   
TUI                         499     21655            -5.5   1151.7   16247   

                           profit_change                  ceo  \
Walmart                             -7.2  C. Douglas McMillon   
State Grid                          -6.2              Kou Wei   
Sinopec Group                      -65.0            Wang Yupu   
China National Petroleum           -73.7        Zhang Jianhua   
Toyota Motor                       -12.3          Akio Toyoda   
...                                  ...                  ...   
National Grid                      160.2       John Pettigrew   
Telecom Italia                       NaN      Flavio Cattaneo   
New China Life Insurance           -45.6             Wan Feng   
Wm. Morrison Supermarkets           20.4       David T. Potts   
TUI                                195.5    Friedrich Joussen   

                                                  industry  \
Walmart                              General Merchandisers   
State Grid                                       Utilities   
Sinopec Group                           Petroleum Refining   
China National Petroleum                Petroleum Refining   
Toyota Motor                      Motor Vehicles and Parts   
...                                                    ...   
National Grid                                    Utilities   
Telecom Italia                          Telecommunications   
New China Life Insurance   Insurance: Life, Health (stock)   
Wm. Morrison Supermarkets             Food and Drug Stores   
TUI                                        Travel Services   

                                           sector  previous_rank  country  \
Walmart                                 Retailing            1.0      USA   
State Grid                                 Energy            2.0    China   
Sinopec Group                              Energy            4.0    China   
China National Petroleum                   Energy            3.0    China   
Toyota Motor               Motor Vehicles & Parts            8.0    Japan   
...                                           ...            ...      ...   
National Grid                              Energy          471.0  Britain   
Telecom Italia                 Telecommunications          404.0    Italy   
New China Life Insurance               Financials          427.0    China   
Wm. Morrison Supermarkets      Food & Drug Stores          437.0  Britain   
TUI                             Business Services          467.0  Germany   

                                 hq_location                       website  \
Walmart                      Bentonville, AR        http://www.walmart.com   
State Grid                    Beijing, China        http://www.sgcc.com.cn   
Sinopec Group                 Beijing, China        http://www.sinopec.com   
China National Petroleum      Beijing, China        http://www.cnpc.com.cn   
Toyota Motor                   Toyota, Japan  http://www.toyota-global.com   
...                                      ...                           ...   
National Grid                London, Britain   http://www.nationalgrid.com   
Telecom Italia 

In [40]:
f500['rank_change'] = rank_change # Attach the rank_change column to the f500 dataframe
f500

rank  revenues  revenue_change  profits  \
Walmart                            1    485873             0.8  13643.0   
State Grid                         2    315199            -4.4   9571.3   
Sinopec Group                      3    267518            -9.1   1257.9   
China National Petroleum           4    262573           -12.3   1867.5   
Toyota Motor                       5    254694             7.7  16899.3   
...                              ...       ...             ...      ...   
Teva Pharmaceutical Industries   496     21903            11.5    329.0   
New China Life Insurance         497     21796           -13.3    743.9   
Wm. Morrison Supermarkets        498     21741           -11.3    406.4   
TUI                              499     21655            -5.5   1151.7   
AutoNation                       500     21609             3.6    430.5   

                                assets  profit_change                  ceo  \
Walmart                         198825           -7.2  C. Douglas McMillon   
State Grid                      489838           -6.2              Kou Wei   
Sinopec Group                   310726          -65.0            Wang Yupu   
China National Petroleum        585619          -73.7        Zhang Jianhua   
Toyota Motor                    437575          -12.3          Akio Toyoda   
...                                ...            ...                  ...   
Teva Pharmaceutical Industries   92890          -79.3    Yitzhak Peterburg   
New China Life Insurance        100609          -45.6             Wan Feng   
Wm. Morrison Supermarkets        11630           20.4       David T. Potts   
TUI                              16247          195.5    Friedrich Joussen   
AutoNation                       10060           -2.7   Michael J. Jackson   

                                                       industry  \
Walmart                                   General Merchandisers   
State Grid                                            Utilities   
Sinopec Group                                Petroleum Refining   
China National Petroleum                     Petroleum Refining   
Toyota Motor                           Motor Vehicles and Parts   
...                                                         ...   
Teva Pharmaceutical Industries                  Pharmaceuticals   
New China Life Insurance        Insurance: Life, Health (stock)   
Wm. Morrison Supermarkets                  Food and Drug Stores   
TUI                                             Travel Services   
AutoNation                                  Specialty Retailers   

                                                sector  previous_rank  \
Walmart                                      Retailing            1.0   
State Grid                                      Energy            2.0   
Sinopec Group                                   Energy            4.0   
China National Petroleum                        Energy            3.0   
Toyota Motor                    Motor Vehicles & Parts            8.0   
...                                                ...            ...   
Teva Pharmaceutical Industries             Health Care            NaN   
New China Life Insurance                    Financials          427.0   
Wm. Morrison Supermarkets           Food & Drug Stores          437.0   
TUI                                  Business Services          467.0   
AutoNation                                   Retailing            NaN   

                                country           hq_location  \
Walmart                             USA       Bentonville, AR   
State Grid                        China        Beijing, China   
Sinopec Group                     China        Beijing, China   
China National Petroleum          China        Beijing, China   
Toyota Motor                      Japan         Toyota, Japan   
...                                 ...                   ...   
Teva Pharmaceutical Industries   Israel  Petach Tikva, Is

By inspecting the f500 dataframe, we realize that the number of rows is 500 despite adding the `rank_change` column with less than 500 rows. What happened is that, when adding a new dataframe to an existing pandas dataframe:

- pandas will ignore the ordering of the index axis and align all the index labels whether string or integer
* Pandas will discard any item on the series whose index (whether string or integer) does not match the existing dataframe
* If the series has a smaller number of rows, pandas will fill them with `NaN` values when they are merged with the existing dataframe

#### Combining Boolean Arrays

To answer more complex questions, we need to learn how to combine boolean arrays. We combine boolean arrays using boolean operators. In Python, these boolean operators are **and**, **or**, and **not**. In pandas, the operators are slightly different:

|Pandas Operator | Normal Python Operator | Meaning|
|:---------------| -----------------------:|---------:|
|a & b|	a and b|	True if both a and b are True, else False|
|a \| b|	a or b|	True if either a or b is True|
|~a|	not a|	True if a is False, else False|

Below, we combined boolean arrays to answer the question of companies with large revenues that still incur negative profits (losses)

In [41]:
large_revenue = f500['revenues'] > 100000
negative_profits = f500['profits'] < 0
combined = large_revenue & negative_profits
big_rev_neg_profit = f500[combined][['revenues', 'profits']]
big_rev_neg_profit

revenues  profits
Japan Post Holdings    122990   -267.4
Chevron                107567   -497.0

The above table shows that just two companies(japan Post Holdings & Chevron) in our dataset made profits of above 1 billion and still incured losses

Just like when we use a single boolean array to perform selection, we don't need to use intermediate variables. The first place we can optimize our code is by combining our two boolean arrays in a single line, instead of assigning them to the intermediate `large_revenue` and `negative_profits` variables first. So a code like this will also work perfectly:

- **combined = (f500["revenues"] > 100000) & (f500["profits"] < 0)**

Notice that we used parentheses around each of our boolean comparisons. This is very important — **our boolean operation will fail without parentheses.**

Lastly, instead of assigning the boolean arrays to `combined`, we can insert the comparison directly into our selection and still get the same output:

- **big_rev_neg_profit = f500[(f500["revenues"] > 100000) & (f500["profits"] < 0)]**

In [42]:
brazil_venezuela = f500[(f500['country'] == 'Brazil') | (f500['country'] == 'Venezuela')]
brazil_venezuela

rank  revenues  revenue_change  profits  \
Petrobras                          75     81405           -16.3  -4838.0   
Itau Unibanco Holding             113     66876            21.4   6666.4   
Banco do Brasil                   151     58093           -13.4   2013.8   
Banco Bradesco                    154     57443            31.3   5127.9   
JBS                               191     48825            -0.1    107.7   
Vale                              370     29363            14.7   3982.0   
Mercantil Servicios Financieros   442     24403            50.3   2004.2   
Ultrapar Holdings                 487     22167            -2.3    447.5   

                                 assets  profit_change  \
Petrobras                        246983            NaN   
Itau Unibanco Holding            415972          -13.7   
Banco do Brasil                  426416          -52.3   
Banco Bradesco                   366418           -5.7   
JBS                               31605          -92.3   
Vale                              99014            NaN   
Mercantil Servicios Financieros  148659          -10.5   
Ultrapar Holdings                  7426           -0.8   

                                                       ceo  \
Petrobras                             Pedro Pullen Parente   
Itau Unibanco Holding              Candido Botelho Bracher   
Banco do Brasil                   Paulo Rogerio Caffarelli   
Banco Bradesco                   Luiz Carlos Trabuco Cappi   
JBS                                Wesley Mendonca Batista   
Vale                                     Fabio Schvartsman   
Mercantil Servicios Financieros      Gustavo J. Vollmer A.   
Ultrapar Holdings                          Thilo Mannhardt   

                                                      industry  \
Petrobras                                   Petroleum Refining   
Itau Unibanco Holding            Banks: Commercial and Savings   
Banco do Brasil                  Banks: Commercial and Savings   
Banco Bradesco                   Banks: Commercial and Savings   
JBS                                            Food Production   
Vale                              Mining, Crude-Oil Production   
Mercantil Servicios Financieros  Banks: Commercial and Savings   
Ultrapar Holdings                                       Energy   

                                                    sector  previous_rank  \
Petrobras                                           Energy           58.0   
Itau Unibanco Holding                           Financials          159.0   
Banco do Brasil                                 Financials          115.0   
Banco Bradesco                                  Financials          209.0   
JBS                              Food, Beverages & Tobacco          185.0   
Vale                                                Energy          417.0   
Mercantil Servicios Financieros                 Financials            NaN   
Ultrapar Holdings                                   Energy          474.0   

                                   country             hq_location  \
Petrobras                           Brazil  Rio de Janeiro, Brazil   
Itau Unibanco Holding               Brazil       Sao Paulo, Brazil   
Banco do Brasil                     Brazil        Brasilia, Brazil   
Banco Bradesco                      Brazil          Osasco, Brazil   
JBS                                 Brazil       Sao Paulo, Brazil   
Vale                                Brazil  Rio de Janeiro, Brazil   
Mercantil Servicios Financieros  Venezuela      Caracas, Venezuela   
Ultrapar Holdings                   Brazil       Sao Paulo, Brazil   

                                                       website  \
Petrobras                          http://www.petrobras.com.br   
Itau Unibanco Holding                   http://www.itau.com.br   
Banco do Brasil                           http://www.bb.com.br   
Banco Bradesco                      http://www.bradesco.com.br   
JBS             

The above table shows that there are 8 companies that are headquartered in either brazil or venezuela

In [43]:
tech_outside_usa = f500[(f500['sector'] == 'Technology') & ~ (f500['country'] == 'USA')] # Tech Firms not headquartered in the USA
tech_outside_usa

rank  revenues  revenue_change  profits  \
Samsung Electronics                   15    173957            -2.0  19316.5   
Hon Hai Precision Industry            27    135129            -4.3   4608.8   
Hitachi                               71     84558             1.2   2134.3   
Huawei Investment & Holding           83     78511            24.9   5579.4   
Sony                                 105     70170             3.9    676.4   
Panasonic                            110     67775             6.7   1378.4   
Amer International Group             183     49677             3.9   1199.9   
LG Electronics                       201     47712            -4.6     66.2   
Lenovo Group                         226     43035            -4.2    535.1   
Fujitsu                              237     41620             5.4    816.7   
JD.com                               261     39155            35.7   -573.0   
Mitsubishi Electric                  262     39119             6.9   1942.6   
Pegatron                             296     35891            -6.1    599.6   
Accenture                            305     34798             5.7   4111.9   
Canon                                347     31271            -0.4   1385.0   
China Electronics                    362     30010            -4.8    321.9   
Taiwan Semiconductor Manufacturing   369     29388            10.6  10283.7   
Quanta Computer                      390     27715           -12.7    469.3   
Schneider Electric                   399     27307            -7.6   1935.2   
Nokia                                415     26113            73.4   -847.1   
LM Ericsson                          419     26004           -11.2    200.5   
NEC                                  437     24596             4.5    252.0   
SAP                                  443     24397             5.8   4031.9   
Midea Group                          450     24060             8.5   2210.4   
Flex                                 455     23863            -2.3    319.6   
Compal Electronics                   458     23773           -10.9    252.1   
Alibaba Group Holding                462     23517            47.9   6489.5   
Onex                                 474     22943             3.8   -130.0   
Tencent Holdings                     478     22871            39.7   6185.9   
LG Display                           479     22840            -9.0    781.4   

                                    assets  profit_change  \
Samsung Electronics                 217104           16.8   
Hon Hai Precision Industry           80436           -0.4   
Hitachi                              86742           48.8   
Huawei Investment & Holding          63837           -5.0   
Sony                                158519          -45.1   
Panasonic                            53702            0.2   
Amer International Group             18405            5.2   
LG Electronics                       31348          -39.8   
Lenovo Group                         27186            NaN   
Fujitsu                              28646           13.0   
JD.com                               23077            NaN   
Mitsubishi Electric                  37519            2.1   
Pegatron                             13776          -20.1   
Accenture                            20609           34.7   
Canon                                44062          -23.9   
China Electronics                    36575           82.6   
Taiwan Semiconductor Manufacturing   58535            7.8   
Quanta Computer                      18229          -16.4   
Schneider Electric                   44137           24.0   
Nokia                                47354         -131.0   
LM Ericsson                          31199          -87.5   
NEC                                  24091          -60.1   
SAP                                  46696           18.6   
Midea Group                          24549            9.3   
Flex                                 12593          -28.0   


In [44]:
tech_outside_usa['sector'].count()

30

The above output shows that there are 30 firms in the technology sector that are not headquartered in the USA

In [45]:
japanese_companies = f500[f500['country'] == 'Japan']
japanese_companies = japanese_companies.sort_values('employees', ascending = False) # descending order (highest to lowest)
japanese_companies[['country', 'employees']].head()

# The sort_values method sorts the values of a column selected from a pandas dataframe in ascending order by default

country  employees
Toyota Motor                   Japan     364445
Hitachi                        Japan     303887
Nippon Telegraph & Telephone   Japan     274844
AEON                           Japan     274760
Panasonic                      Japan     257533

The above output shows the top 5 japanese companies with the highest number employees

In [46]:
top_japanese_employees = japanese_companies.loc['Toyota Motor']['employees'] # Selecting a single item from a dataframe using df.loc[]
top_japanese_employees

364445

### Aggregate Calculations

Aggregation is where we apply a statistical operation to groups of our data. We can use the aggregation technique to calculate the average revenue for each unique country in our dataset, we can follow the steps:

- create an empty dictionary called `avg_rev_by_country` that will hold the as key each unique country and as value the average revenue generated by all companies in that country
- create a variable `countries` that will hold an array of the unique countries in the original `f500` dataframe
- loop through each country `c` in the variable `countries` created in step 2 above
- for each iteration create a dtaframe `selected_rows` that select all rows in the original `f500` dataframe where, the `country` column of the original `f500` dataframe equals the unique values in the `countries` array variable created in step 2
- create a variable called `mean` that will hold all the selected unique rows in the `selected_rows` dataframe from step 4 above, select only the `revenues` column and use the `mean` method to find the mean of the revenues for all unique countries
- update the `avg_rev_by_country` dictionary that has its keys as `c` which is the unique countries and holds as value `mean` which is the average revenues of companies in each unique country

In [47]:
avg_rev_by_country = {}
countries = country.unique()
for c in countries:
    selected_rows = f500[f500['country'] == c] 
    mean = selected_rows['revenues'].mean()
    avg_rev_by_country[c] = mean
avg_rev_by_country

{'USA': 64218.371212121216,
 'China': 55397.880733944956,
 'Japan': 53164.03921568627,
 'Germany': 63915.0,
 'Netherlands': 61708.92857142857,
 'Britain': 51588.708333333336,
 'South Korea': 49725.6,
 'Switzerland': 51353.57142857143,
 'France': 55231.793103448275,
 'Taiwan': 46364.666666666664,
 'Singapore': 54454.333333333336,
 'Italy': 51899.57142857143,
 'Russia': 65247.75,
 'Spain': 40600.666666666664,
 'Brazil': 52024.57142857143,
 'Mexico': 54987.5,
 'Luxembourg': 56791.0,
 'India': 39993.0,
 'Malaysia': 49479.0,
 'Thailand': 48719.0,
 'Australia': 33688.71428571428,
 'Belgium': 45905.0,
 'Norway': 45873.0,
 'Canada': 31848.0,
 'Ireland': 32819.5,
 'Indonesia': 36487.0,
 'Denmark': 35464.0,
 'Saudi Arabia': 35421.0,
 'Sweden': 27963.666666666668,
 'Finland': 26113.0,
 'Venezuela': 24403.0,
 'Turkey': 23456.0,
 'U.A.E': 22799.0,
 'Israel': 21903.0}

### CEO's that employ the highest employees in each country


In [48]:
top_employer_by_country = {}
countries = f500['country'].unique()
for c in countries:
    unique_rows = f500[f500["country"] == c] # pick from the country column a country whose value equals a unique country
    sorted_rows = unique_rows.sort_values('employees', ascending = False) # sort the employee column to descending order
    top_employer = sorted_rows.iloc[0] # get the first row from the sorted rows that has the employee for each iteration
    employer_name = top_employer['ceo'] # select the ceo from the top_employer row
    top_employer_by_country[c] = employer_name
top_employer_by_country

{'USA': 'C. Douglas McMillon',
 'China': 'Zhang Jianhua',
 'Japan': 'Akio Toyoda',
 'Germany': 'Matthias Muller',
 'Netherlands': 'John Elkann',
 'Britain': 'Richard J. Cousins',
 'South Korea': 'Oh-Hyun Kwon',
 'Switzerland': 'Ulf Mark Schneider',
 'France': 'Michel Landel',
 'Taiwan': 'Terry Gou',
 'Singapore': 'Michael M. McNamara',
 'Italy': 'Matteo Del Fante',
 'Russia': 'Alexey B. Miller',
 'Spain': 'Jose Antonio Alvarez',
 'Brazil': 'Wesley Mendonca Batista',
 'Mexico': 'Daniel Hajj Aboumrad',
 'Luxembourg': 'Lakshmi N. Mittal',
 'India': 'Arundhati Bhattacharya',
 'Malaysia': 'Wan Zulkiflee Wan Ariffin',
 'Thailand': 'Tevin Vongvanich',
 'Australia': 'Richard J.B. Goyder',
 'Belgium': 'Carlos Brito',
 'Norway': 'Eldar Saetre',
 'Canada': 'Galen G. Weston',
 'Ireland': 'Pierre Nanterme',
 'Indonesia': 'Elia Massa Manik',
 'Denmark': 'Soren Skou',
 'Saudi Arabia': 'Yousef Abdullah Al-Benyan',
 'Sweden': 'Karl-Johan Persson',
 'Finland': 'Rajeev Suri',
 'Venezuela': 'Gustavo J. Vo

### Interpretation of the above output

The output above shows the highest employers in each country. From the output, we can see that in the USA the highest employer is 'C. Douglas McMillon', in china, the CEO that employs the highest number of workers is 'Zhang Jianhua', while in Japan 'Akio Toyoda', has the highest number of employees.

In [49]:
roa = f500['profits'] / f500['assets']
f500['roa'] = roa

top_roa_by_country = {}
unique_country = f500['country'].unique()
for s in unique_country:
    selected_country = f500[f500['country'] == s]
    mean_roa_of_each_country = selected_country['roa'].mean()
    top_roa_by_country[s] = mean_roa_of_each_country
top_roa_by_country
# converting the dictionary values to descending order (from highest roa to lowest)
import operator
sorted_dict_to_list = sorted(top_roa_by_country.items(), key = operator.itemgetter(1), reverse = True)
# Above, operator.itemgetter(1) takes the value of the key which is at the index 1. The output is a list
# Converting output back to a dictionary
sorted_dict = dict(sorted_dict_to_list)
sorted_dict

{'Sweden': 0.08221458155602634,
 'Indonesia': 0.06662714627485021,
 'Ireland': 0.06562718919368628,
 'Saudi Arabia': 0.05630437098320492,
 'Taiwan': 0.055179687540293364,
 'India': 0.052046119448705054,
 'USA': 0.04801502254153256,
 'Turkey': 0.045682117618876514,
 'Thailand': 0.04300951097852411,
 'Switzerland': 0.03541744950265122,
 'Spain': 0.03494376239674667,
 'Britain': 0.03323462083704367,
 'South Korea': 0.031013987315335795,
 'Russia': 0.030825278637977564,
 'Malaysia': 0.030424149619410087,
 'Japan': 0.028300750865267538,
 'France': 0.025244389757711476,
 'Netherlands': 0.02472697244132069,
 'Canada': 0.024060044099073027,
 'Luxembourg': 0.02367517500199622,
 'Singapore': 0.023280730181668378,
 'Germany': 0.020418818119650715,
 'Brazil': 0.01700577785188523,
 'China': 0.01653725168231649,
 'Venezuela': 0.013481861172212917,
 'U.A.E': 0.010282209330432682,
 'Belgium': 0.004802984739589985,
 'Italy': 0.004226145455572219,
 'Israel': 0.003541823662396383,
 'Australia': -0.010091

The above code shows the country with the **Average** highest Returns on Assets

**Note: (ROA is a business-specific metric which indicates a companies ability to make profit using their available assets)** 

Thus, on the **average**, switzerland has the highest returns on assests, followed by indonesia and then ireland at third position.

The countries with the least **Average** returns on assets are Norway, Denmark and Mexico

In [51]:
roa = f500['profits'] / f500['assets']
f500['roa'] = roa

top_roa_by_country = {}
unique_country = f500['country'].unique()
for s in unique_country:
    selected_country = f500[f500['country'] == s]
    srt = selected_country.sort_values('roa', ascending = False)
    lock = srt.iloc[0][-1]
    top_roa_by_country[s] = lock
top_roa_by_country
# converting the dictionary values to descending order (from highest roa to lowest)
import operator
sorted_dict_to_list = sorted(top_roa_by_country.items(), key = operator.itemgetter(1), reverse = True)
# Above, operator.itemgetter(1) takes the value of the key which is at the index 1. The output is a list
# Converting output back to a dictionary
sorted_dict = dict(sorted_dict_to_list)
sorted_dict

{'USA': 0.23695526264984115,
 'Sweden': 0.2052523171987642,
 'Ireland': 0.19951962734727544,
 'Taiwan': 0.1756846331254805,
 'Spain': 0.16436353346224591,
 'Netherlands': 0.16363791485368143,
 'Switzerland': 0.12855480167704902,
 'Britain': 0.1233215891143239,
 'China': 0.10858552169639095,
 'Japan': 0.10509800758247964,
 'India': 0.0992805173807599,
 'Canada': 0.09700097529258778,
 'France': 0.0913989940655188,
 'South Korea': 0.08897348736089616,
 'Germany': 0.0863435840328936,
 'Indonesia': 0.06662714627485021,
 'Brazil': 0.06026124427686507,
 'Saudi Arabia': 0.05630437098320492,
 'Russia': 0.05129660754088191,
 'Turkey': 0.045682117618876514,
 'Thailand': 0.04300951097852411,
 'Malaysia': 0.030424149619410087,
 'Italy': 0.026911636045494316,
 'Singapore': 0.02625297040397494,
 'Luxembourg': 0.02367517500199622,
 'Venezuela': 0.013481861172212917,
 'U.A.E': 0.010282209330432682,
 'Australia': 0.009753606956980038,
 'Mexico': 0.006293249949017742,
 'Belgium': 0.004802984739589985,
 '

The above code shows the country with the **highest Returns on Assets**

Thus, the United States has the highest returns on assests, followed by sweden and then ireland at third position.

The countries with the least returns on assets are finland, norway and denmark.

### Difference between the returns on assets and the average returns on assests

The immediate table above shows the returns on assests for countries, starting from the country with the highest returns to the country with the smallest returns. The metric used is the **ROA** as it is, without further manipulations.

What the code for the above output does is that, it groups together all the companies that are located in a single country and isolates the company with the highest **ROA** for each unique country. After which it displays the country and the Highest **ROA** for that country.

The code that checks for average does the same thing but with some slight differences. It computes the average **ROA** for all the companies that are in a unique country and presents that average value alongside the country. That is, it gets the **ROA** for all companies within a country and divides the **total ROA** by the number of companies within the unique country.


So while this code takes into consideration all the respective **ROA's** for all companies within a unique country, the code without average does not take all the other companies **ROA** in account, rather it simply prints out the highest **ROA** for each unique country.

That is why when reading the results from the **Average ROA** Switzerland appears first followed by indonesia and then ireland (Having smaller companies within it, thus the average produced a higher result). This is because the number of companies within these countries is relatively smaller than the number of companies in countries like the USA (who have more companies within it, thus the average produced a lower result) topping the rankings when calculating **ROA without Average**